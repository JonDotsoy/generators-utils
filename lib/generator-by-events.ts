import EventEmitter from "events";

const endSymbol = Symbol("end");
type endSymbol = typeof endSymbol;

const isEndSymbol = (value: unknown): value is endSymbol => value === endSymbol

export interface initCallback<T> {
    (arg: {
        push: (value: T) => void
        error: (error: any) => void
        done: () => void
    }): void;
}

export function generatorByEvents<T>(init: initCallback<T>): AsyncGenerator<T> {
    const events = new EventEmitter();
    const refError: { error?: any } = {};
    let finished = false;
    const changes: (T | endSymbol)[] = []

    const data = (value: T | endSymbol) => {
        const emitted = events.emit("data", value);
        if (!emitted) {
            changes.push(value);
        }
    }

    init({
        push: (v: T) => data(v),
        done: () => data(endSymbol),
        error: (e: any) => {
            refError.error = e;
            data(endSymbol);
        },
    })

    return {
        next: async () => {
            if (finished) return { value: undefined, done: true }
            if (Object.hasOwnProperty.call(refError, 'error')) {
                finished = true
                throw refError.error
            }
            if (changes.length === 0) {
                return new Promise<{ value: T, done: boolean }>((resolve, reject) => {
                    const handler = (snap: T | endSymbol) => {
                        if (Object.hasOwnProperty.call(refError, 'error')) {
                            finished = true
                            reject(refError.error)
                            return
                        }
                        if (isEndSymbol(snap)) {
                            resolve({ value: undefined!, done: true })
                            return
                        }
                        resolve({ value: snap, done: false });
                    }
                    events.once('data', handler);
                });
            }
            const value = changes.shift()!;
            if (isEndSymbol(value)) {
                return { value: undefined, done: true }
            }
            return { value, done: false };
        },
        return: async (value?: any) => {
            data(endSymbol);
            return { value, done: true };
        },
        throw: async (error: any) => {
            data(endSymbol);
            return { value: undefined, done: true };
        },
        [Symbol.asyncIterator]() {
            return this;
        },
    }
}
