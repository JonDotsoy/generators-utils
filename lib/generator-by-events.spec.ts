import { initCallback, generatorByEvents } from "./generator-by-events"

describe('generatorByEvents', () => {

    it('sample', async () => {
        const cb = jest.fn()

        const init: initCallback<number> = (a) => {
            a.push(1);
            a.push(2);
            a.push(3);
            setTimeout(() => {
                a.push(4);
                a.push(5);
                a.push(6);
                a.push(7);
                a.push(8);
                a.push(9);
            }, 300);
            setTimeout(() => {
                a.done()
            }, 400);
            
        }

        for await (const hi of generatorByEvents<number>(init)) {
            cb(hi)
        }

        expect(cb.mock.calls.length).toBe(9)
        expect(cb.mock.calls).toMatchSnapshot()

    })

    it('', async () => {
        const init: initCallback<number> = (a) => {
            a.push(1);
            a.push(2);
            a.push(3);
            setTimeout(() => {
                a.push(4);
                a.push(5);
            }, 300);

            setTimeout(() => {
                a.error(new Error('error'));
                a.push(6);
                a.push(7);
            }, 350);
            
            setTimeout(() => {
                a.done()
            }, 400);
            
        }

        try {
            for await (const hi of generatorByEvents<number>(init)) {
                console.log({ hi })
            }
        } catch (e) {
            console.log({ e })
        }
    })

})
