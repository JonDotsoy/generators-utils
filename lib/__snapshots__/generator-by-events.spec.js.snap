// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatorByEvents sample 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
  Array [
    4,
  ],
  Array [
    5,
  ],
  Array [
    6,
  ],
  Array [
    7,
  ],
  Array [
    8,
  ],
  Array [
    9,
  ],
]
`;
